src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker.cs

```diff
@@ -132,72 +132,67 @@ public async Task<ImmutableArray<ProjectInfo>> LoadAsync(CancellationToken cance
 
-                        if (!processedPaths.Contains(projectPath))
-                        {
-                            results.AddRange(projectInfos);
-                        }
+                        continue;
                     }
 
-                    return results.ToImmutable();
+                    var projectFileInfos = await LoadProjectInfosFromPathAsync(absoluteProjectPath, _requestedProjectOptions, cancellationToken).ConfigureAwait(false);
+
+                    results.AddRange(projectFileInfos);
                 }
-                finally
+
+                foreach (var (projectPath, projectInfos) in _pathToDiscoveredProjectInfosMap)
                 {
-                    _buildManager.EndBatchBuild();
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    if (!processedPaths.Contains(projectPath))
+                    {
+                        results.AddRange(projectInfos);
+                    }
                 }
+
+                return results.ToImmutable();
             }
 
             private async Task<ImmutableArray<ProjectFileInfo>> LoadProjectFileInfosAsync(string projectPath, DiagnosticReportingOptions reportingOptions, CancellationToken cancellationToken)
             {
-                if (!_projectFileLoaderRegistry.TryGetLoaderFromProjectPath(projectPath, reportingOptions.OnLoaderFailure, out var loader))
+                if (!_projectFileExtensionRegistry.TryGetLanguageNameFromProjectPath(projectPath, reportingOptions.OnLoaderFailure, out var languageName))
                 {
                     return ImmutableArray<ProjectFileInfo>.Empty; // Failure should already be reported.
                 }
 
+                var (buildHost, buildHostPreferredKind) = await _buildHostProcessManager.GetBuildHostAsync(projectPath, cancellationToken).ConfigureAwait(false);
                 var projectFile = await DoOperationAndReportProgressAsync(
                     ProjectLoadOperation.Evaluate,
                     projectPath,
                     targetFramework: null,
-                    () => loader.LoadProjectFileAsync(projectPath, _buildManager, cancellationToken)
+                    () => buildHost.LoadProjectFileAsync(projectPath, languageName, cancellationToken)
                 ).ConfigureAwait(false);
 
                 // If there were any failures during load, we won't be able to build the project. So, bail early with an empty project.
-                if (projectFile.Log.HasFailure)
+                var diagnosticItems = await projectFile.GetDiagnosticLogItemsAsync(cancellationToken).ConfigureAwait(false);
+                if (diagnosticItems.Any(d => d.Kind == WorkspaceDiagnosticKind.Failure))
                 {
-                    _diagnosticReporter.Report(projectFile.Log);
+                    _diagnosticReporter.Report(diagnosticItems);
 
+                    // TODO: correct language type!
```

Comment:
yeah, seems wonky.  what impact does this have?

Reply:
Ooh oops, sorry that should have gotten cleaned up since we now have languageName available.

Reply:
Fixed now.